File: go/utility/prepare_auth.go
package utility

import (
	"strings"

	vs "github.com/voxgig-sdk/tiktok-engagement-bot-sdk/go/utility/struct"

	"github.com/voxgig-sdk/tiktok-engagement-bot-sdk/go/core"
)

const credName = "PHPSESSID"
const cookieHeader = "cookie"
const optionApikey = "apikey"
const notFound = "__NOTFOUND__"

func cookieSet(headers map[string]any, value any) {
	kept := []string{}

	if existing, ok := headers[cookieHeader].(string); ok && existing != "" {
		for _, part := range strings.Split(existing, ";") {
			piece := strings.TrimSpace(part)
			if piece == "" || piece == credName ||
				strings.HasPrefix(piece, credName+"=") {
				continue
			}
			kept = append(kept, piece)
		}
	}

	if value != nil {
		valStr, _ := value.(string)
		kept = append(kept, credName+"="+valStr)
	}

	if len(kept) == 0 {
		delete(headers, cookieHeader)
	} else {
		headers[cookieHeader] = strings.Join(kept, "; ")
	}
}

func prepareAuthUtil(ctx *core.Context) (*core.Spec, error) {
	spec := ctx.Spec
	if spec == nil {
		return nil, ctx.MakeError("auth_no_spec",
			"Expected context spec property to be defined.")
	}

	headers := spec.Headers
	options := ctx.Client.OptionsMap()

	// Public APIs that need no auth omit the options.auth block entirely.
	if options["auth"] == nil {
		cookieSet(headers, nil)
		return spec, nil
	}

	apikey := vs.GetProp(options, optionApikey, notFound)

	skip := false
	if apikey == nil {
		skip = true
	} else if apikeyStr, ok := apikey.(string); ok &&
		(apikeyStr == notFound || apikeyStr == "") {
		skip = true
	}

	if skip {
		cookieSet(headers, nil)
	} else {
		apikeyVal := ""
		if av, ok := apikey.(string); ok {
			apikeyVal = av
		}
		cookieSet(headers, apikeyVal)
	}

	return spec, nil
}

File: php/utility/PrepareAuth.php
<?php
declare(strict_types=1);

// TiktokEngagementBot SDK utility: prepare_auth

class TiktokEngagementBotPrepareAuth
{
    private const COOKIE_AUTH = 'PHPSESSID';
    private const HEADER_COOKIE = 'cookie';
    private const OPTION_APIKEY = 'apikey';
    private const NOT_FOUND = '__NOTFOUND__';

    private static function applyCookie(array &$headers, ?string $value): void
    {
        $kept = [];
        $existing = $headers[self::HEADER_COOKIE] ?? '';

        if (is_string($existing) && '' !== $existing) {
            foreach (explode(';', $existing) as $part) {
                $piece = trim($part);
                if ('' === $piece || $piece === self::COOKIE_AUTH
                    || str_starts_with($piece, self::COOKIE_AUTH . '=')) {
                    continue;
                }
                $kept[] = $piece;
            }
        }

        if (null !== $value) {
            $kept[] = self::COOKIE_AUTH . '=' . $value;
        }

        if ([] === $kept) {
            unset($headers[self::HEADER_COOKIE]);
        } else {
            $headers[self::HEADER_COOKIE] = implode('; ', $kept);
        }
    }

    public static function call(TiktokEngagementBotContext $ctx): array
    {
        $spec = $ctx->spec;
        if (!$spec) {
            return [null, $ctx->make_error('auth_no_spec', 'Expected context spec property to be defined.')];
        }

        $headers = &$spec->headers;
        $options = $ctx->client->options_map();

        // Public APIs that need no auth omit the options.auth block entirely.
        if (!isset($options['auth']) || $options['auth'] === null) {
            self::applyCookie($headers, null);
            return [$spec, null];
        }

        $apikey = \Voxgig\Struct\Struct::getprop($options, self::OPTION_APIKEY, self::NOT_FOUND);

        $missing = (is_string($apikey) && ($apikey === self::NOT_FOUND || $apikey === ''))
            || $apikey === null;

        if ($missing) {
            self::applyCookie($headers, null);
        } else {
            // One `Cookie:` header holds every cookie, separated by '; '.
            self::applyCookie($headers, is_string($apikey) ? $apikey : '');
        }

        return [$spec, null];
    }
}
